onnxruntime/core/common/threadpool.cc

```diff
@@ -57,6 +57,29 @@ void ThreadPool::ParallelFor(int32_t total, std::function<void(int32_t)> fn) {
   barrier.Wait();
 }
 
+void ThreadPool::BatchParallelFor(int32_t total, std::function<void(int32_t)> fn, int32_t batch_size) {
+  if (total <= 0)
+    return;
+
+  if (total == 1 || batch_size <= 1) {
+    fn(0);
+    return;
+  }
+
+  if (batch_size == total) {
+    ParallelFor(total, fn);
+    return;
+  }
+
+  ParallelFor(batch_size, [&](int batch_index) {
+    int start = batch_index * total / batch_size;
+    int end = (batch_index + 1) * total / batch_size;
+    for (int i = start; i < end; i++) {
```

Comment:
Does this assume total is evenly divisible by batch size? 
Should it do something like std::min(((batch_index + 1) * total / batch_size), total) to make sure it doesn't run past the end?

Reply:
there is no need to calculate as the current value of `start` and `end` will make sure every item be called exactly once, and even in best effort.